src/components/formBuilder/preview/FormPreview.tsx

```
@@ -71,6 +72,20 @@ const FormPreview: React.FC<FormPreviewProps> = ({
 
         unwrapTemplateExpressions(draft);
 
+        if (typeof draftSchema.properties === "object") {
+          for (const value of Object.values(draftSchema.properties).filter(
+            (value) =>
+              typeof value === "object" && value.$ref === databaseSchema.$id
+          )) {
+            // @ts-expect-error -- value is a schema object
+            value.type = "string";
+            // @ts-expect-error -- value is a schema object
+            value.enum = "Select...";
+            // @ts-expect-error -- value is a schema object
+            delete value.$ref;
```

Can you cast the array to `as Schema[]` on line 79 inside the parenthesis, so we get typing and don't have to suppress errors here? Also, `enum` is an array, not a string, on the `Schema` type. TS highlights this error if you cast the array.

---

Thanks for the suggestion, updated.

Regarding `enum`, our RJSF field treats string enums in a specific way: it converts them to an array (this matches the `Schema` requirement), sets as default value, and disables the dropdown. Pls check the updated comments in the code (I'm happy to change the copy if it doesn't read clear).

---

Yes.
1. `@var` comes directly from the form builder;
2. Disabling the dropdown. That is assigned via component props (`disabled`) and doesn't come from schema. In case of `@var` (and now database) we want to show just one option in the select and make the control not available. Expressions are not resolved in preview, so it doesn't make much sense to open a Dropdown assigned to a variable. With database a disabled dropdown allows to use a simple SelectWidget rather than DatabaseWidget, so it doesn't load database options (I'm not sure this would work at all tbh).

---

Ok, that makes sense. FYI, json schema has a `readOnly: boolean` field that we could probably use here as opposed to the separate `disabled` component prop. I'm not sure if that would cause other issues or end up being even more complicated or whatever 🤷‍♂️

---

In [the section on `readOnly` and `writeOnly`](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01#section-10.3), the json schema spec even says:

>These keywords can be used to assist in user interface instance generation.

---

Good suggestion. Unfortunately `readOnly` property doesn't make a dropdown disabled it works for all the fields except those with dropdown).
I created a ticket to track the issue https://github.com/pixiebrix/pixiebrix-extension/issues/4414